### PR TITLE
[funimation] establish session before logging in

### LINKED ISSUE #10629: [Funimation] Unable to extract players data
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2016.09.11.1*. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2016.09.11.1**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add `-v` flag to **your command line** you run youtube-dl with, copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):
```
C:\Users\Martin\Desktop\test>youtube-dl.exe "http://www.funimation.com/shows/seiyus-life/videos/official/recording" -u user -p password -v
[debug] System config: []
[debug] User config: []
[debug] Command-line args: ['http://www.funimation.com/shows/seiyus-life/videos/official/recording', '-u', 'PRIVATE', '-p', 'PRIVATE', '-v']
[debug] Encodings: locale cp1252, fs mbcs, out cp850, pref cp1252
[debug] youtube-dl version 2016.09.11.1
[debug] Python version 3.4.4 - Windows-10-10.0.14393
[debug] exe versions: ffmpeg N-81338-g6612d04, ffprobe N-81338-g6612d04, rtmpdump 2.4
[debug] Proxy map: {}
[Funimation] Logging in as user
[Funimation] recording: Downloading webpage
ERROR: Unable to extract players data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpj47q005f\build\youtube_dl\YoutubeDL.py", line 691, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpj47q005f\build\youtube_dl\extractor\common.py", line 347, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpj47q005f\build\youtube_dl\extractor\funimation.py", line 157, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpj47q005f\build\youtube_dl\extractor\common.py", line 650, in _search_regex
youtube_dl.utils.RegexNotFoundError: Unable to extract players data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

### Description of your *issue*, suggested solution and other information

Seems to be a recent issue, last time I downloaded something about 8 days ago everything still worked.
Other sites (I tried Crunchyroll) do not seem to be affected.

### LINKED ISSUE #10639: [funimation] Unable to log in
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2016.09.11.1*. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2016.09.11.1**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add `-v` flag to **your command line** you run youtube-dl with, copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

_Note: this error can be reproduced on both the latest release and Git HEAD. The log output is identical for both._

```
$ python -m youtube_dl --verbose http://www.funimation.com/shows/love-live-sunshine/videos/official/aye-aye-my-friend\?watch\=sub
[debug] System config: []
[debug] User config: [u'--netrc', u'--cookies', u'/home/parent5446/.cache/youtube-dl-cookies
.txt', u'--prefer-free-formats']                                                           
[debug] Command-line args: [u'--verbose', u'http://www.funimation.com/shows/love-live-sunshi
ne/videos/official/aye-aye-my-friend?watch=sub']                                           
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2016.09.11.1
[debug] Git HEAD: 8414c2d
[debug] Python version 2.7.12+ - Linux-4.6.0-1-amd64-x86_64-with-debian-stretch-sid
[debug] exe versions: ffmpeg 3.1.2-1, ffprobe 3.1.2-1, rtmpdump 2.4
[debug] Proxy map: {}
[Funimation] Logging in as tylerromeo@gmail.com
ERROR: Unable to log in; please report this issue on https://yt-dl.org/bug . Make sure you a
re using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to cal
l youtube-dl with the --verbose flag and include its complete output.                      
Traceback (most recent call last):
  File "youtube_dl/YoutubeDL.py", line 692, in extract_info
    ie_result = ie.extract(url)
  File "youtube_dl/extractor/common.py", line 346, in extract
    self.initialize()
  File "youtube_dl/extractor/common.py", line 340, in initialize
    self._real_initialize()
  File "youtube_dl/extractor/funimation.py", line 113, in _real_initialize
    self._login()
  File "youtube_dl/extractor/funimation.py", line 110, in _login
    raise ExtractorError('Unable to log in')
ExtractorError: Unable to log in; please report this issue on https://yt-dl.org/bug . Make s
ure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be su
re to call youtube-dl with the --verbose flag and include its complete output.
```

---

### Description of your *issue*, suggested solution and other information

Funimation is having login issues. It may be related to changes in their site. Trying to investigate now, but no guarantees.

### PR DESCRIPTION
**Pull request by Parent5446**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added an initial HTTP GET request with the login URL before attempting
to log in. This allows Funimation to set any session cookies it
requires. This request is necessary for Funimation to render the login
page or to accept login requests.

Fixes #10639
Fixes #10629

### CODE REVIEW

File: youtube_dl/extractor/funimation.py [hunk 1]
```diff
@@ -87,26 +87,40 @@ def _login(self):
...
+
+        # Log in
         login_request = sanitized_Request(self._LOGIN_URL, data, headers={
             'User-Agent': user_agent,
             'Content-Type': 'application/x-www-form-urlencoded'
         })
         login_page = self._download_webpage(
             login_request, None, 'Logging in as %s' % username)
+
+        # Check for errors
```
Comment (dstftw) [inline youtube_dl/extractor/funimation.py]: Actually, it's an opposite. It's a check for successful login.

File: youtube_dl/extractor/funimation.py [hunk 2]
```diff
@@ -87,26 +87,40 @@ def _login(self):
...
         user_agent = self._extract_cloudflare_session_ua(self._LOGIN_URL)
         if not user_agent:
             user_agent = 'Mozilla/5.0 (Windows NT 5.2; WOW64; rv:42.0) Gecko/20100101 Firefox/42.0'
+
+        # Request home page to establish initial session and cookies
+        initial_request = sanitized_Request(self._LOGIN_URL, headers={
+            'User-Agent': user_agent,
+            'Content-Type': 'application/x-www-form-urlencoded'
+        })
+        self._download_webpage(initial_request, None, 'Starting session with Funimation')
```
Comment (dstftw) [inline youtube_dl/extractor/funimation.py]: `headers` now can be passed right in `_download_webpage`. Also there is already the same dict in `login_request` that may be extracted and reused.

Comment (Parent5446) [inline youtube_dl/extractor/funimation.py]: Done with amended commit.

File: youtube_dl/extractor/funimation.py (hunk 1 above)
Comment (Parent5446) [inline youtube_dl/extractor/funimation.py]: Done with amended commit.

**Comment by dstftw** [comment]
Current code from master works fine for me at the moment. Can you try?

**Comment by MartinWeingaertner** [comment]
Current release build 2016.09.15 works fine for my referenced issue, but can't double check with the old version right now.

**Comment by dstftw** [comment]
2016.09.15 has no funimation changes.

**Comment by MartinWeingaertner** [comment]
And yet I'm currently downloading the video that didn't work for me before with that version. So, a regression on Funis part?

**Comment by Parent5446** [comment]
I'm just going to guess Funimation undid whatever it did before?

**Comment by Parent5446** [comment]
Just verified on my computer as well. This should no longer be needed.

**Pull request closed without merging**